com.unity.netcode.gameobjects/Components/NetworkAnimator.cs

```diff
@@ -120,104 +210,113 @@ public bool SetTrigger(int key)
 
             private void SerializeTriggerParameters<T>(BufferSerializer<T> serializer) where T : IReaderWriter
             {
-                int paramCount = serializer.IsReader ? 0 : TriggerParameters.Count;
+                int paramCount;
+                if (serializer.IsReader)
+                {
+                    paramCount = 0;
+                    TriggerParameters = new HashSet<int>(paramCount);
+                }
+                else
+                {
+                    paramCount = TriggerParameters.Count;
+                }
+
                 serializer.SerializeValue(ref paramCount);
 
-                var paramArray = serializer.IsReader ? new int[paramCount] : TriggerParameters.ToArray();
-                for (int i = 0; i < paramCount; i++)
+                foreach (var thisParamId in TriggerParameters)
                 {
-                    var paramId = serializer.IsReader ? 0 : paramArray[i];
+                    var paramId = serializer.IsReader ? 0 : thisParamId;
                     serializer.SerializeValue(ref paramId);
 
                     if (serializer.IsReader)
                     {
-                        paramArray[i] = paramId;
+                        TriggerParameters.Add(paramId);
                     }
                 }
-
-                if (serializer.IsReader)
-                {
-                    TriggerParameters = new HashSet<int>(paramArray);
-                }
             }
 
             private void SerializeBoolParameters<T>(BufferSerializer<T> serializer) where T : IReaderWriter
             {
-                int paramCount = serializer.IsReader ? 0 : BoolParameters.Count;
+                int paramCount = serializer.IsReader ? 0 : BoolParamArray.Length;
                 serializer.SerializeValue(ref paramCount);
 
-                var paramArray = serializer.IsReader ? new KeyValuePair<int, bool>[paramCount] : BoolParameters.ToArray();
+                if (BoolParamArray.Length != paramCount)
+                {
+                    BoolParamArray = new KeyValuePair<int, bool>[paramCount];
+                }
+
                 for (int paramIndex = 0; paramIndex < paramCount; paramIndex++)
                 {
-                    var paramId = serializer.IsReader ? 0 : paramArray[paramIndex].Key;
+                    var paramId = serializer.IsReader ? 0 : BoolParamArray[paramIndex].Key;
```

Comment:
Like, rather than have to make a custom serializer for `KeyValuePair<int, float>`, etc., could we not store these as 2 arrays `FloatKeys<int>` and `FloatValues<float>` and then let our new serialization code do this for us?